design/cluster-external-secret-spec.md

```diff
@@ -0,0 +1,117 @@
+```yaml
+---
+title: Adding Cluster External Secrets
+version: v1alpha1
+authors: Daniel "ADustyOldMuffin" Hix
+creation-date: 2020-09-01
+status: draft
+---
+```
+
+# Adding Cluster External Secrets
+
+## Table of Contents
+
+<!-- toc -->
+- [Adding Cluster External Secrets](#adding-cluster-external-secrets)
+  - [Table of Contents](#table-of-contents)
+  - [Summary](#summary)
+  - [Motivation](#motivation)
+    - [Goals](#goals)
+    - [Non-Goals](#non-goals)
+  - [Proposal](#proposal)
+    - [User Stories](#user-stories)
+    - [API](#api)
+    - [Behavior](#behavior)
+    - [Drawbacks](#drawbacks)
+    - [Acceptance Criteria](#acceptance-criteria)
+  - [Alternatives](#alternatives)
+<!-- /toc -->
+
+
+## Summary
+This would provide a way to template an `ExternalSecret` and based on a namespace selector in the CRD it would generate `ExternalSecret`s in matching namespaces.
+
+## Motivation
+It's a pain point to have to create a `Secret`/`ExternalSecret` in every namespace where it's needed, and this would provide a way to to do this easily. Another motivation is the possible creation of a Kubernetes secret provider which would provide an `ExternalSecret` from a secret already located in cluster. This in combination with that provider could provide a way to sync a secret across namespaces with a single provider call.
+
+### Goals
+To provide a way to deploy multiple `ExternalSecret`s with a single CRD.
+
+### Non-Goals
+Lower provider calls based on the created `ExternalSecrets` or manage their sync.
+
+## Proposal
+
+### User Stories
+As an ESO User I would like to create the same `Secret`/`ExternalSecret` in multiple namespaces.
+
+### API
+``` yaml
+apiVersion: external-secrets.io/v1alpha1
+kind: ClusterExternalSecret
+metadata:
+  name: testing-secret
+spec:
+  # The selector used to select the namespaces to deploy to
+  namespaceSelector:
+    matchLabels:
+      foo: bar
+  # This spec is the exact same spec as an ExternalSecret, and is used as a template
+  externalSecretSpec:
+    refreshInterval: "15s"
+    secretStoreRef:
+      name: some-provider
+      kind: ClusterSecretStore
+    target:
+      name: my-cool-new-secret
+      creationPolicy: Owner
+    data:
+    - secretKey: my-cool-new-secret-key
+      remoteRef:
+        key: test
+        property: foo
+```
+
+### Behavior
+When created the controller will find namespaces via a label selector, and matching namespaces will then have an `ExternalSecret` deployed to them.
+
+Edge cases are, 
+
+1. namespaces being labeled after creation - currently handled via a re-queue interval, but the interval is a tad high and the changes won't take place right away.
```

Comment:
We can consider moving the re-queue interval to the CES spec or configure it with a cli flag.

Reply:
I added a refresh interval just like external secrets which changes when the controller is requeued.